text/0026-sassafras-consensus.md

```diff
@@ -0,0 +1,1079 @@
+
+In both scenarios, the signature within the `SlotClaim` is verified using
+a `VrfSignData` constructed as specified by paragraph 6.5.
+
+Given `claim` an instance of `SlotClaim`:
+
+```rust
+    public_key = AUTHORITIES[claim.authority_index];
+
+    vrf_verify(public_key, sign_data, claim.signature);
+```
+
+If signature verification fails then the claim is not legit.
+
+### 6.6.1. Primary Claim Method Verification
+
+This verification is performed to confirm ticket ownership and is performed
+utilizing the second `VrfOutput` contained within the `SlotClaim` `signature`.
+
+By using the `VrfOutput` object together with the associated expected `VrfInput`
+the verifier should be able to reconstruct the `revealed_pub` key committed in
+the `TicketBody`. If there is a mismatch, the claim is not legit.
+
+```rust
+    revealed_vrf_input = vrf_input_from_items(
+        domain: BYTES("sassafras-revealed-v1.0"),
+        data: [
+            curr_epoch.randomness,
+            BYTES(curr_epoch.epoch_index),
+            BYTES(ticket_body.attempt_index)
+        ]
+    );
+
+    reveled_vrf_output = claim.signature.outputs[1];
+
+    revealed_seed = vrf_bytes(32, revealed_vrf_input, revealed_vrf_output);
+    revealed_pub = ed25519_secret_from_seed(revealed_seed).public();
+
+    assert(revealed_pub == ticket_body.revealed_pub);
+```
+
+##### 6.6.1.1. (Optional) Ephemeral Key Signature Check
+
+If the `erased_signature` element within the `SlotClaim` is present the
+`erased_pub` key is used to verify it.
+
+The signed challenge is generated with identical steps as outlined in section
+6.5.1.1.
+
+```rust
+    challenge = sign_data.transcript.challenge();
+    result = ed25519_verify(ticket_body.erased_pub, challenge, claim.erased_signature);
+
+    assert(result == true);
+```
+
+#### 6.6.2. Secondary Claim Method Verification
+
+If the slot doesn't have any associated ticket then the validator index contained in
+the claim should match the one given by the rule outlined in section [6.4.2](642-fallback-assignment).
+
+### 6.7. Randomness Accumulator
+
+The first `VrfOutput` which ships with the block's `SlotClaim` `signature`
+is mandatory and must be used as the entropy source for the randomness which
+gets accumulated on-chain **after** block processing.
+
+Given `claim` the instance of `SlotClaim` within the block header, and
+`accumulator` the current value for the current epoch randomness accumulator,
+the `accumulator` value is updated as follows:
+
+```rust
+    randomness_vrf_input = vrf_input_from_items(
+        domain: BYTES("sassafras-randomness-v1.0"),
+        data: [
+            curr_epoch.randomness,
+            BYTES(curr_epoch.epoch_index),
+            BYTES(slot)
+        ]
+    );
+
+    randomness_vrf_output = claim.signature.outputs[0];
+
+    randomness = vrf_bytes(32, randomness_vrf_input, randomness_vrf_output);
+
+    accumulator = BLAKE2(32, CONCAT(accumulator, randomness));
+```
```

Comment:
`accumulator` just accumulates forever here, right?

We should not reset `accumulator = ""` anyware except maybe in genesis (and maybe the genesis hash or whatever is a better choice)

Reply:
Never resets. On genesis, I'll put there genesis hash 👍